python/ray/dag/dag_node_operation.py

```diff
@@ -343,28 +453,42 @@ def _generate_actor_to_execution_schedule(
 
     visited_nodes = set()
 
-    # Use topological sort algorithm to generate the execution schedule. Each iteration
-    # pops a candidate node from `actor_to_candidates` and each DAG node consists of
-    # three operations: READ, COMPUTE, and WRITE.
-    for _ in range(len(graph) * 3):
-        # The function `_select_next_nodes` will pop a candidate node from
-        # `actor_to_candidates` and return a list of nodes that can be executed
-        # in the next step. If multiple nodes are returned, only the NCCL write
-        # node is popped in this iteration.
+    # Use topological sort algorithm to generate the execution schedule.
+    while True:
+        # Select a list of nodes to be executed. There are three cases:
+        # 1. If a selected node is not an NCCL operation, only itself is returned.
+        # 2. If a selected node is an NCCL write operation, the corresponding NCCL
+        #    read operations are also returned.
+        # 3. If a selected node is an NCCL compute operation, all the nodes in its
+        #    collective group are returned.
+        # In cases 1 and 3, all the selected nodes are ready. In case 2, the NCCL
+        # write node is ready, while the NCCL read nodes are not ready until their
+        # in-degrees are updated.
         nodes = _select_next_nodes(actor_to_candidates, graph)
+        if nodes is None:
+            break
+        # Filter out the visited nodes.
+        nodes = [node for node in nodes if node not in visited_nodes]
+        # Add the selected nodes to the execution schedule.
         for node in nodes:
-            if node in visited_nodes:
-                continue
             actor_to_execution_schedule[node.actor_handle].append(node.operation)
             visited_nodes.add(node)
+        # Update the in-degree of the downstream nodes.
+        for node in nodes:
             for out_node_task_idx, out_node_type in node.out_edges:
                 out_node = graph[out_node_task_idx][out_node_type]
                 out_node.in_edges.remove((node.task_idx, node.operation.type))
-                if out_node.in_degree == 0:
+                if out_node.in_degree == 0 and out_node not in visited_nodes:
```

Comment:
Can we replace `in_degree == 0` with `is_ready` everywhere?

Reply:
No. For a collective node, it is different for cases `in_degree == 0` and `is_ready == True`.